# 1. استكشاف البيانات الشعرية

# Exploring Arabic Poem Comprehensive Dataset

In [1]:
import pandas as pd
import numpy as np
poem_df = pd.read_csv("C:\\Users\\snmuj\\Documents\\NLP--Course1\\Arabic_Poem Comprehensive_Dataset.csv")
print("Shape:", poem_df.shape)
print("Index:", poem_df.index)
poem_df.head()

Shape: (1831770, 9)
Index: RangeIndex(start=0, stop=1831770, step=1)


Unnamed: 0        العصر             الشاعر          الديوان القافية  \
0           0  قبل الإسلام  عمرو بنِ قُمَيئَة  الديوان الرئيسي       د   
1           1  قبل الإسلام  عمرو بنِ قُمَيئَة  الديوان الرئيسي       د   
2           2  قبل الإسلام  عمرو بنِ قُمَيئَة  الديوان الرئيسي       د   
3           3  قبل الإسلام  عمرو بنِ قُمَيئَة  الديوان الرئيسي       د   
4           4  قبل الإسلام  عمرو بنِ قُمَيئَة  الديوان الرئيسي       د   

    البحر                           الشطر الايسر  \
0  الطويل   وَأَن تَجمَعا شَملي وَتَنتَظِرا غَدا   
1  الطويل  وَلا سُرعَتي يَوماً بِسابِقَةِ الرَدى   
2  الطويل  وَتَستَوجِبا مَنّاً عَلَيَّ وَتُحمَدا   
3  الطويل    تُؤامِرُني سِرّاً لِأَصرِمَ مَرثَدا   
4  الطويل   وَأَفرَعَ في لَومي مِراراً وَأَصعَدا   

                             الشطر الايمن  \
0   خَليلَيَّ لا تَستَعجِلا أَن تَزَوَّدا   
1     فَما لَبَثٌ يَوماً بِسابِقٍ مَغنَمٍ   
2  وَإِن تُنظِراني اليَومَ أَقضِ لُبانَةً   
3      لَعَمرُكَ ما نَفسٌ بِجِدٍ رَشيدَةٍ   
4    وَإِن ظَهَرَت مِنهُ قَوارِصُ جَمَّةٌ   

                                               البيت  
0  خَليلَيَّ لا تَستَعجِلا أَن تَزَوَّدا    وَأَن...  
1  فَما لَبَثٌ يَوماً بِسابِقٍ مَغنَمٍ    وَلا سُ...  
2  وَإِن تُنظِراني اليَومَ أَقضِ لُبانَةً    وَتَ...  
3  لَعَمرُكَ ما نَفسٌ بِجِدٍ رَشيدَةٍ    تُؤامِرُ...  
4  وَإِن ظَهَرَت مِنهُ قَوارِصُ جَمَّةٌ    وَأَفر...

In [2]:
poem_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1831770 entries, 0 to 1831769
Data columns (total 9 columns):
 #   Column        Dtype 
---  ------        ----- 
 0   Unnamed: 0    int64 
 1   العصر         object
 2   الشاعر        object
 3   الديوان       object
 4   القافية       object
 5   البحر         object
 6   الشطر الايسر  object
 7   الشطر الايمن  object
 8   البيت         object
dtypes: int64(1), object(8)
memory usage: 125.8+ MB


In [3]:
poem_df.isna().sum().sum()

160074

In [4]:
poem_df['العصر'].isna().sum()

0

In [5]:
poem_df['الشاعر'].isna().sum()

0

In [6]:
poem_df['الديوان'].isna().sum()

27

In [7]:
poem_df['القافية'].isna().sum()

9760

In [8]:
poem_df['البحر'].isna().sum()

92334

In [9]:
poem_df['الشطر الايسر'].isna().sum()

57910

In [10]:
poem_df['الشطر الايمن'].isna().sum()

43

In [11]:
poem_df['البيت'].isna().sum()

0

In [12]:
df = poem_df.dropna()
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1684668 entries, 0 to 1831769
Data columns (total 9 columns):
 #   Column        Non-Null Count    Dtype 
---  ------        --------------    ----- 
 0   Unnamed: 0    1684668 non-null  int64 
 1   العصر         1684668 non-null  object
 2   الشاعر        1684668 non-null  object
 3   الديوان       1684668 non-null  object
 4   القافية       1684668 non-null  object
 5   البحر         1684668 non-null  object
 6   الشطر الايسر  1684668 non-null  object
 7   الشطر الايمن  1684668 non-null  object
 8   البيت         1684668 non-null  object
dtypes: int64(1), object(8)
memory usage: 128.5+ MB


In [13]:
df['العصر'].nunique()

12

In [14]:
df['العصر'].unique()

array(['قبل الإسلام', 'المخضرمين', 'العباسي', 'الأموي', 'المملوكي',
       'المغرب والأندلس', 'بين الدولتين', 'الفاطمي', 'الأيوبي', 'الحديث',
       'الإسلامي', 'العثماني'], dtype=object)

In [15]:
df['الشاعر'].nunique()

3366

In [16]:
df['الشاعر'].unique()

array(['عمرو بنِ قُمَيئَة', 'السُليَك بن السَلَكَة', 'عمرو بن بَرّاقَة',
       ..., 'الربيع بن البارع الزوزني', 'ممويه الكرخي',
       'أصبهدوست الديلمي'], dtype=object)

In [17]:
df['الديوان'].nunique()

749

In [18]:
df['القافية'].nunique()

35

In [19]:
df['القافية'].unique()

array(['د', 'ح', 'م', 'ل', 'ف', 'ب', 'ر', 'ي', 'ء', 'ت', 'ن', 'ج', 'ذ',
       'ز', 'س', 'ش', 'ض', 'ع', 'ق', 'ك', 'ه', 'ط', 'ث', 'و', 'ا', 'خ',
       'غ', 'ص', 'ظ', 'ى', 'هـ', 'هن', 'لا', 'ؤ', 'طن'], dtype=object)

In [20]:
df['البحر'].nunique()

23

In [21]:
df['البحر'].unique()

array(['الطويل', 'المنسرح', 'المتقارب', 'الخفيف', 'الكامل', 'السريع',
       'الوافر', 'البسيط', 'الرجز', 'الرمل', 'المجتث', 'المديد', 'الهزج',
       'المتدارك', 'المقتضب', 'المضارع', 'شعر التفعيلة', 'الدوبيت',
       'موشح', 'السلسلة', 'المواليا', 'شعر حر', 'عامي'], dtype=object)

# 2. البيانات بعد استبعاد الخلايا الفارغة

# Output after remvoing the balnk rows 
***
### Dataset
- **What is the dataset about ?**

* عدد المداخل: 1831770
* عدد البيانات المفقودة 160074
* في الديوان 27
* في القافية 9760
* في البحر 92334
* في الشطر الأيسر 57910
* في الشطر الأيمن 43

### البيانات بعد فلترة الخلايا الفارغة
* إزالة البيانات المفقودة
* عدد الأسطر بعد الإزالة 1684668 

### معلومات أولية عن بعض المتغيرات
* عدد العصور 12
* عدد الشعراء 3366
* عدد الدواين 749
* عدد القوافي 35
* عدد البحور 23
### <div class="alert alert-block alert-success"> مهام المعالجة بعد تصفية الباينات المفقودة هو إجراء حساب تكرار الكلمة تكرار  المستند العكسي ثم استخراج الموضوعات وقد تزيد لاحقا </div>

In [22]:
df.to_csv('clean_Arabic_Poem Comprehensive_Dataset.csv', encoding="utf-8-sig")

# 3. تصنيف البيانات وفق بعض العصور

In [23]:
preislamic = pd.read_csv("clean_Arabic_Poem Comprehensive_Dataset.csv")

In [24]:
preislamic.loc[preislamic['العصر'] == ('قبل الإسلام')]

Unnamed: 0.1  Unnamed: 0        العصر                 الشاعر  \
0                   0           0  قبل الإسلام      عمرو بنِ قُمَيئَة   
1                   1           1  قبل الإسلام      عمرو بنِ قُمَيئَة   
2                   2           2  قبل الإسلام      عمرو بنِ قُمَيئَة   
3                   3           3  قبل الإسلام      عمرو بنِ قُمَيئَة   
4                   4           4  قبل الإسلام      عمرو بنِ قُمَيئَة   
...               ...         ...          ...                    ...   
1681060       1826323     1826323  قبل الإسلام            درهم بن زيد   
1681061       1826324     1826324  قبل الإسلام            درهم بن زيد   
1683929       1830596     1830596  قبل الإسلام  سعدة بنت فريد الأسدية   
1683930       1830597     1830597  قبل الإسلام  سعدة بنت فريد الأسدية   
1683931       1830598     1830598  قبل الإسلام  سعدة بنت فريد الأسدية   

                             الديوان القافية    البحر  \
0                    الديوان الرئيسي       د   الطويل   
1                    الديوان الرئيسي       د   الطويل   
2                    الديوان الرئيسي       د   الطويل   
3                    الديوان الرئيسي       د   الطويل   
4                    الديوان الرئيسي       د   الطويل   
...                              ...     ...      ...   
1681060              الديوان الرئيسي       ف  المنسرح   
1681061              الديوان الرئيسي       ف  المنسرح   
1683929  ديوان سعدة بنت فريد الأسدية       م   الطويل   
1683930  ديوان سعدة بنت فريد الأسدية       م   الطويل   
1683931  ديوان سعدة بنت فريد الأسدية       م   الطويل   

                                  الشطر الايسر  \
0         وَأَن تَجمَعا شَملي وَتَنتَظِرا غَدا   
1        وَلا سُرعَتي يَوماً بِسابِقَةِ الرَدى   
2        وَتَستَوجِبا مَنّاً عَلَيَّ وَتُحمَدا   
3          تُؤامِرُني سِرّاً لِأَصرِمَ مَرثَدا   
4         وَأَفرَعَ في لَومي مِراراً وَأَصعَدا   
...                                        ...   
1681060              وسابغاتٌ كأنَّها النُطَفُ   
1681061          بها نُفوسُ الكُماةِ تُختَطَفُ   
1683929             عليك بتخدين النساء الكرائم   
1683930           بريش الدنابي لا بريش القوادم   
1683931               وللشرف العاديّ بان وهادم   

                                   الشطر الايمن  \
0         خَليلَيَّ لا تَستَعجِلا أَن تَزَوَّدا   
1           فَما لَبَثٌ يَوماً بِسابِقٍ مَغنَمٍ   
2        وَإِن تُنظِراني اليَومَ أَقضِ لُبانَةً   
3            لَعَمرُكَ ما نَفسٌ بِجِدٍ رَشيدَةٍ   
4          وَإِن ظَهَرَت مِنهُ قَوارِصُ جَمَّةٌ   
...                                         ...   
1681060             البيضُ حصنٌ لهم إذا فزِعُوا   
1681061           والبيضُ قد ثُلِّمَت مضارِبُها   
1683929              عليك بأنقاض العراق فقد علت   
1683930             لعمري لقد راش ابن سعدة نفسه   
1683931                 بنى لك معروف بناء هدمته   

                                                     البيت  
0        خَليلَيَّ لا تَستَعجِلا أَن تَزَوَّدا    وَأَن...  
1        فَما لَبَثٌ يَوماً بِسابِقٍ مَغنَمٍ    وَلا سُ...  
2        وَإِن تُنظِراني اليَومَ أَقضِ لُبانَةً    وَتَ...  
3        لَعَمرُكَ ما نَفسٌ بِجِدٍ رَشيدَةٍ    تُؤامِرُ...  
4        وَإِن ظَهَرَت مِنهُ قَوارِصُ جَمَّةٌ    وَأَفر...  
...                                                    ...  
1681060  البيضُ حصنٌ لهم إذا فزِعُوا    وسابغاتٌ كأنَّه...  
1681061  والبيضُ قد ثُلِّمَت مضارِبُها    بها نُفوسُ ال...  
1683929  عليك بأنقاض العراق فقد علت    عليك بتخدين النس...  
1683930  لعمري لقد راش ابن سعدة نفسه    بريش الدنابي لا...  
1683931  بنى لك معروف بناء هدمته    وللشرف العاديّ بان ...  

[21324 rows x 10 columns]

In [25]:
preislamic.loc[preislamic['العصر'] == ('قبل الإسلام')].to_csv('clean_preislamic-poems.csv', encoding="utf-8-sig")

In [26]:
abbasid = pd.read_csv("clean_Arabic_Poem Comprehensive_Dataset.csv")

In [27]:
abbasid.loc[abbasid['العصر'] == ('العباسي')]

Unnamed: 0.1  Unnamed: 0    العصر                     الشاعر  \
387               396         396  العباسي                المُتَنَبّي   
388               397         397  العباسي                المُتَنَبّي   
389               398         398  العباسي                المُتَنَبّي   
390               399         399  العباسي                المُتَنَبّي   
391               400         400  العباسي                المُتَنَبّي   
...               ...         ...      ...                        ...   
1684651       1831753     1831753  العباسي  أبو علي الدامغاني القومسي   
1684652       1831754     1831754  العباسي  أبو علي الدامغاني القومسي   
1684653       1831755     1831755  العباسي  أبو علي الدامغاني القومسي   
1684654       1831756     1831756  العباسي  أبو علي الدامغاني القومسي   
1684655       1831757     1831757  العباسي  أبو علي الدامغاني القومسي   

                           الديوان القافية   البحر  \
387                الديوان الرئيسي       ء  الكامل   
388                الديوان الرئيسي       ء  الكامل   
389                الديوان الرئيسي       ء  الكامل   
390                الديوان الرئيسي       ء  الكامل   
391                الديوان الرئيسي       ء  الكامل   
...                            ...     ...     ...   
1684651  أبو علي الدامغاني القومسي       س  الطويل   
1684652  أبو علي الدامغاني القومسي       س  الطويل   
1684653  أبو علي الدامغاني القومسي       س  الطويل   
1684654  أبو علي الدامغاني القومسي       س  الطويل   
1684655  أبو علي الدامغاني القومسي       س  الطويل   

                                 الشطر الايسر  \
387      وَأَحَقُّ مِنكَ بِجَفنِهِ وَبِمائِهِ   
388       قَسَماً بِهِ وَبِحُسنِهِ وَبَهائِهِ   
389       إِنَّ المَلامَةَ فيهِ مِن أَعدائِهِ   
390       دَع ما بَراكَ ضَعُفتَ عَن إِخفائِهِ   
391         وَأَرى بِطَرفٍ لا يَرى بِسَوائِهِ   
...                                       ...   
1684651             رشىً لهم من ترهات البسابس   
1684652          حلال اتساعاً في فنون المقايس   
1684653        وما سجلوا أيضاً بها في الحبايس   
1684654           جفوني وأنسى بالوحوش الكوانس   
1684655          عراص المعالي كالطلول الدوارس   

                                     الشطر الايمن  \
387            القَلبُ أَعلَمُ يا عَذولُ بِدائِهِ   
388      فَوَمَن أُحِبُّ لَأَعصِيَنَّكَ في الهَوى   
389           أَأُحِبُّهُ وَأُحِبُّ فيهِ مَلامَةً   
390      عَجِبَ الوُشاةُ مِنَ اللُحاةِ وَقَولِهِم   
391        ما الخِلُّ إِلّا مَن أَوَدُّ بِقَلبِهِ   
...                                           ...   
1684651                يعدون ما دون البتيكات وضحا   
1684652            بها حللوا عين الحرام وحرموا ال   
1684653            كما غصبوا الأملاك معشوقة الورى   
1684654             فيا وحشتي منهم إذا اكتحلت بهم   
1684655                مضى الرؤساء الأولون واصبحت   

                                                     البيت  
387      القَلبُ أَعلَمُ يا عَذولُ بِدائِهِ    وَأَحَقّ...  
388      فَوَمَن أُحِبُّ لَأَعصِيَنَّكَ في الهَوى    قَ...  
389      أَأُحِبُّهُ وَأُحِبُّ فيهِ مَلامَةً    إِنَّ ا...  
390      عَجِبَ الوُشاةُ مِنَ اللُحاةِ وَقَولِهِم    دَ...  
391      ما الخِلُّ إِلّا مَن أَوَدُّ بِقَلبِهِ    وَأَ...  
...                                                    ...  
1684651  يعدون ما دون البتيكات وضحا    رشىً لهم من ترها...  
1684652  بها حللوا عين الحرام وحرموا ال    حلال اتساعاً...  
1684653  كما غصبوا الأملاك معشوقة الورى    وما سجلوا أي...  
1684654  فيا وحشتي منهم إذا اكتحلت بهم    جفوني وأنسى ب...  
1684655  مضى الرؤساء الأولون واصبحت    عراص المعالي كال...  

[231513 rows x 10 columns]

In [28]:
abbasid.loc[abbasid['العصر'] == ('العباسي')].to_csv('clean_abbasid-poems.csv', encoding="utf-8-sig")

In [29]:
andalus = pd.read_csv("clean_Arabic_Poem Comprehensive_Dataset.csv")

In [30]:
andalus.loc[andalus['العصر'] == ('المغرب والأندلس')]

Unnamed: 0.1  Unnamed: 0            العصر  \
74511           74928       74928  المغرب والأندلس   
74512           74929       74929  المغرب والأندلس   
74513           74930       74930  المغرب والأندلس   
74514           74931       74931  المغرب والأندلس   
74515           74932       74932  المغرب والأندلس   
...               ...         ...              ...   
1684531       1831633     1831633  المغرب والأندلس   
1684532       1831634     1831634  المغرب والأندلس   
1684533       1831635     1831635  المغرب والأندلس   
1684534       1831636     1831636  المغرب والأندلس   
1684535       1831637     1831637  المغرب والأندلس   

                              الشاعر                      الديوان القافية  \
74511             أبو اسحاق الألبيري              الديوان الرئيسي       ت   
74512             أبو اسحاق الألبيري              الديوان الرئيسي       ت   
74513             أبو اسحاق الألبيري              الديوان الرئيسي       ت   
74514             أبو اسحاق الألبيري              الديوان الرئيسي       ت   
74515             أبو اسحاق الألبيري              الديوان الرئيسي       ت   
...                              ...                          ...     ...   
1684531  الحسن بن هود المرسي الجذامي  الحسن بن هود المرسي الجذامي       ل   
1684532  الحسن بن هود المرسي الجذامي  الحسن بن هود المرسي الجذامي       ل   
1684533  الحسن بن هود المرسي الجذامي  الحسن بن هود المرسي الجذامي       ل   
1684534  الحسن بن هود المرسي الجذامي  الحسن بن هود المرسي الجذامي       ل   
1684535  الحسن بن هود المرسي الجذامي  الحسن بن هود المرسي الجذامي       ل   

          البحر                      الشطر الايسر  \
74511    الوافر  وَتَنحِتُ جِسمَكَ الساعاتُ نَحتا   
74512    الوافر   أَلا يا صاحِ أَنتَ أُريدُ أَنتا   
74513    الوافر  أَبَتَّ طَلاقَها الأَكياسُ بَتّا   
74514    الوافر  بِها حَتّى إِذا مِتَّ اِنتَبَهنا   
74515    الوافر     مَتى لا تَرعَوي عَنها وَحَتّى   
...         ...                               ...   
1684531   الرمل                   أنا علم أنا جهل   
1684532   الرمل                    أنا جزء أنا كل   
1684533   الرمل                    أنا عقد أنا حل   
1684534   الرمل                   أنا هجر أنا وصل   
1684535   الرمل                أنا حَزْنٌ أنا سهل   

                             الشطر الايمن  \
74511    تَفُتُّ فُؤادَكَ الأَيّامُ فَتّا   
74512     وَتَدعوكَ المَنونُ دُعاءَ صِدقٍ   
74513    أَراكَ تُحِبُّ عِرساً ذاتَ غَدرٍ   
74514    تَنامُ الدَهرَ وَيحَكَ في غَطيطٍ   
74515      فَكَم ذا أَنتَ مَخدوعٌ وَحَتّى   
...                                   ...   
1684531                    أنا سر أنا جهر   
1684532                   أنا حرب أنا سلم   
1684533                   أنا قبض أنا بسط   
1684534                   أنا بعد أنا قرب   
1684535                    أنا حلو أنا مر   

                                                     البيت  
74511    تَفُتُّ فُؤادَكَ الأَيّامُ فَتّا    وَتَنحِتُ ...  
74512    وَتَدعوكَ المَنونُ دُعاءَ صِدقٍ    أَلا يا صاح...  
74513    أَراكَ تُحِبُّ عِرساً ذاتَ غَدرٍ    أَبَتَّ طَ...  
74514    تَنامُ الدَهرَ وَيحَكَ في غَطيطٍ    بِها حَتّى...  
74515    فَكَم ذا أَنتَ مَخدوعٌ وَحَتّى    مَتى لا تَرع...  
...                                                    ...  
1684531                  أنا سر أنا جهر    أنا علم أنا جهل  
1684532                  أنا حرب أنا سلم    أنا جزء أنا كل  
1684533                  أنا قبض أنا بسط    أنا عقد أنا حل  
1684534                 أنا بعد أنا قرب    أنا هجر أنا وصل  
1684535               أنا حلو أنا مر    أنا حَزْنٌ أنا سهل  

[107088 rows x 10 columns]

In [31]:
andalus.loc[andalus['العصر'] == ('المغرب والأندلس')].to_csv('clean_andalus-poems.csv', encoding="utf-8-sig")

In [32]:
otman = pd.read_csv("clean_Arabic_Poem Comprehensive_Dataset.csv")

In [33]:
otman.loc[otman['العصر'] == ('العثماني')]

Unnamed: 0.1  Unnamed: 0     العصر             الشاعر  \
374618         383525      383525  العثماني         بنت الشحنة   
374619         383526      383526  العثماني         بنت الشحنة   
374620         383527      383527  العثماني         بنت الشحنة   
374621         383528      383528  العثماني         بنت الشحنة   
374622         383529      383529  العثماني         بنت الشحنة   
...               ...         ...       ...                ...   
1684389       1831491     1831491  العثماني  عبد الكريم الفكون   
1684390       1831492     1831492  العثماني  عبد الكريم الفكون   
1684391       1831493     1831493  العثماني  عبد الكريم الفكون   
1684392       1831494     1831494  العثماني  عبد الكريم الفكون   
1684393       1831495     1831495  العثماني  عبد الكريم الفكون   

                   الديوان القافية   البحر                    الشطر الايسر  \
374618     الديوان الرئيسي       ن  البسيط       وجلت فينا بجد ليسٍ بالحسن   
374619     الديوان الرئيسي       ن  البسيط  أوليتني في الورى حزناً على حزن   
374620     الديوان الرئيسي       ن  البسيط      أخذت مني محب الدين من وطني   
374621     الديوان الرئيسي       ن  البسيط   يا ليتني قبل ذا أدرجت في كفني   
374622     الديوان الرئيسي       ن  البسيط      فليت بعد عفيف الدين لم أكن   
...                    ...     ...     ...                             ...   
1684389  عبد الكريم الفكون       ض  الطويل         بقافية بالعجز عنها يعرض   
1684390  عبد الكريم الفكون       ض  الطويل  إذا عاصف التثريب في الوجه ينفض   
1684391  عبد الكريم الفكون       ض  الطويل    على من لرجس الشرك ينفي ويدحض   
1684392  عبد الكريم الفكون       ض  الطويل    وما دام داعي الله للكفر يعرض   
1684393  عبد الكريم الفكون       ض  الطويل          لكل أمور الله تجر مفوض   

                           الشطر الايمن  \
374618   يا بين بالغت في الأشجان والمحن   
374619    أضرمت نار فؤادي والحشاءَ معاً   
374620        أغلقت باب علوم ثم باب هدى   
374621      قد مات في غربة والشام مسكنه   
374622       وقد فقدت عفيف الدين واأسفي   
...                                 ...   
1684389           تنبأته من سوقه لقصيدة   
1684390  فما أقبح التقصير والجهل بالفتى   
1684391           وصل إلهي كل وقت وساعة   
1684392      وآله والأصحاب ما هبت الصبا   
1684393       وأنصاره والتابعين لهم ومن   

                                                     البيت  
374618   يا بين بالغت في الأشجان والمحن    وجلت فينا بج...  
374619   أضرمت نار فؤادي والحشاءَ معاً    أوليتني في ال...  
374620   أغلقت باب علوم ثم باب هدى    أخذت مني محب الدي...  
374621   قد مات في غربة والشام مسكنه    يا ليتني قبل ذا...  
374622   وقد فقدت عفيف الدين واأسفي    فليت بعد عفيف ال...  
...                                                    ...  
1684389   تنبأته من سوقه لقصيدة    بقافية بالعجز عنها يعرض  
1684390  فما أقبح التقصير والجهل بالفتى    إذا عاصف الت...  
1684391  وصل إلهي كل وقت وساعة    على من لرجس الشرك ينف...  
1684392  وآله والأصحاب ما هبت الصبا    وما دام داعي الل...  
1684393  وأنصاره والتابعين لهم ومن    لكل أمور الله تجر...  

[151146 rows x 10 columns]

In [34]:
otman.loc[otman['العصر'] == ('العثماني')].to_csv('clean_otman-poems.csv', encoding="utf-8-sig")

In [35]:
modern = pd.read_csv("clean_Arabic_Poem Comprehensive_Dataset.csv")

In [36]:
modern.loc[modern['العصر'] == ('الحديث')]

Unnamed: 0.1  Unnamed: 0   العصر     الشاعر          الديوان القافية  \
195118         198186      198186  الحديث  أحمد شوقي  الديوان الرئيسي       ء   
195119         198187      198187  الحديث  أحمد شوقي  الديوان الرئيسي       ء   
195120         198188      198188  الحديث  أحمد شوقي  الديوان الرئيسي       ء   
195121         198189      198189  الحديث  أحمد شوقي  الديوان الرئيسي       ء   
195122         198190      198190  الحديث  أحمد شوقي  الديوان الرئيسي       ء   
...               ...         ...     ...        ...              ...     ...   
1684663       1831765     1831765  الحديث  شهاب غانم        شهاب غانم       ن   
1684664       1831766     1831766  الحديث  شهاب غانم        شهاب غانم       ن   
1684665       1831767     1831767  الحديث  شهاب غانم        شهاب غانم       ن   
1684666       1831768     1831768  الحديث  شهاب غانم        شهاب غانم       ن   
1684667       1831769     1831769  الحديث  شهاب غانم        شهاب غانم       ن   

          البحر                       الشطر الايسر  \
195118   الخفيف    وَحَداها بِمَن تُقِلُّ الرَجاءُ   
195119   الخفيف  ها سَماءً قَد أَكبَرَتها السَماءُ   
195120   الخفيف     ضِ شِباكاً تَمُدُّها الدَأماءُ   
195121   الخفيف      تَتَدَجّى كَأَنَّها الظَلماءُ   
195122   الخفيف      لُ وَهاجَت حُماتَها الهَيجاءُ   
...         ...                                ...   
1684663  الخفيف                وأحلى قصيدة تَتَغنى   
1684664  الخفيف            كحلم يغشى الجفون الوسنى   
1684665  الخفيف            يتداعى وجداً ويخفق حسنا   
1684666  الخفيف             يدك الحدود سجناً فسجنا   
1684667  الخفيف             آه لو أدرك الغرام لجنا   

                               الشطر الايمن  \
195118   هَمَّتِ الفُلكُ وَاِحتَواها الماءُ   
195119   ضَرَبَ البَحرُ ذو العُبابِ حَوالَي   
195120   وَرَأى المارِقونَ مِن شَرَكِ الأَر   
195121        وَجِبالاً مَوائِجاً في جِبالٍ   
195122    وَدَوِيّاً كَما تَأَهَّبَتِ الخَي   
...                                     ...   
1684663    هي أغلى ما أنشأ اللَّه في الدنيا   
1684664            هي أغرودة الأغاريد تنساب   
1684665                  هي شلال بهجة وبهاء   
1684666         هي حلم الهوى ومنطلقي الباقي   
1684667             هي حبي العاتي وكل غرامي   

                                                     البيت  
195118   هَمَّتِ الفُلكُ وَاِحتَواها الماءُ    وَحَداها...  
195119   ضَرَبَ البَحرُ ذو العُبابِ حَوالَي    ها سَماء...  
195120   وَرَأى المارِقونَ مِن شَرَكِ الأَر    ضِ شِباك...  
195121   وَجِبالاً مَوائِجاً في جِبالٍ    تَتَدَجّى كَأ...  
195122   وَدَوِيّاً كَما تَأَهَّبَتِ الخَي    لُ وَهاجَ...  
...                                                    ...  
1684663  هي أغلى ما أنشأ اللَّه في الدنيا    وأحلى قصيد...  
1684664  هي أغرودة الأغاريد تنساب    كحلم يغشى الجفون ا...  
1684665      هي شلال بهجة وبهاء    يتداعى وجداً ويخفق حسنا  
1684666  هي حلم الهوى ومنطلقي الباقي    يدك الحدود سجنا...  
1684667  هي حبي العاتي وكل غرامي    آه لو أدرك الغرام لجنا  

[669688 rows x 10 columns]

In [37]:
modern.loc[modern['العصر'] == ('الحديث')].to_csv('clean_modern-poems.csv', encoding="utf-8-sig")

# 4. إزالة كلمات الإيقاف

# عدد الكلمات لكل بيت 

# عدد البحور مقابل العصور، وعدد القوافي مقابل العصور، وعدد الكلمات لكل بيت بناء على العصور، إلخ

In [38]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
import re
import pickle 
import time
import nltk
from nltk import Text
from nltk.tokenize import regexp_tokenize
from nltk.tokenize import word_tokenize  
from nltk.tokenize import sent_tokenize 
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
from nltk.stem import PorterStemmer
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression 
from sklearn.naive_bayes import MultinomialNB
from sklearn.multiclass import OneVsRestClassifier
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import GridSearchCV
from sklearn.pipeline import make_pipeline
import pandas as pd
import numpy as np
df = pd.read_csv('clean_Arabic_Poem Comprehensive_Dataset.csv', encoding="utf-8-sig")

In [39]:
df['word_count'] = df['البيت'].apply(lambda x: len(str(x).split(" ")))
df[['البيت','word_count']].head()

البيت  word_count
0  خَليلَيَّ لا تَستَعجِلا أَن تَزَوَّدا    وَأَن...          13
1  فَما لَبَثٌ يَوماً بِسابِقٍ مَغنَمٍ    وَلا سُ...          13
2  وَإِن تُنظِراني اليَومَ أَقضِ لُبانَةً    وَتَ...          12
3  لَعَمرُكَ ما نَفسٌ بِجِدٍ رَشيدَةٍ    تُؤامِرُ...          12
4  وَإِن ظَهَرَت مِنهُ قَوارِصُ جَمَّةٌ    وَأَفر...          13

In [40]:
def avg_word(sentence):
  words = sentence.split()
  return (sum(len(word) for word in words)/len(words))
df['avg_word'] = df['البيت'].apply(lambda x: avg_word(x))
df[['البيت','avg_word']].head()

البيت  avg_word
0  خَليلَيَّ لا تَستَعجِلا أَن تَزَوَّدا    وَأَن...  6.500000
1  فَما لَبَثٌ يَوماً بِسابِقٍ مَغنَمٍ    وَلا سُ...  6.400000
2  وَإِن تُنظِراني اليَومَ أَقضِ لُبانَةً    وَتَ...  7.555556
3  لَعَمرُكَ ما نَفسٌ بِجِدٍ رَشيدَةٍ    تُؤامِرُ...  6.888889
4  وَإِن ظَهَرَت مِنهُ قَوارِصُ جَمَّةٌ    وَأَفر...  6.400000

In [41]:
from nltk.corpus import stopwords
stop = stopwords.words('function-ar')
df['stopwords'] = df['البيت'].apply(lambda x: len([x for x in x.split() if x in stop]))
df[['البيت','stopwords']].head()

البيت  stopwords
0  خَليلَيَّ لا تَستَعجِلا أَن تَزَوَّدا    وَأَن...          1
1  فَما لَبَثٌ يَوماً بِسابِقٍ مَغنَمٍ    وَلا سُ...          0
2  وَإِن تُنظِراني اليَومَ أَقضِ لُبانَةً    وَتَ...          0
3  لَعَمرُكَ ما نَفسٌ بِجِدٍ رَشيدَةٍ    تُؤامِرُ...          1
4  وَإِن ظَهَرَت مِنهُ قَوارِصُ جَمَّةٌ    وَأَفر...          1

In [42]:
df['punctuations'] = df['البيت'].apply(lambda x: len([x for x in x.split() if x.startswith('.,:')]))
df[['البيت','punctuations']].head()

البيت  punctuations
0  خَليلَيَّ لا تَستَعجِلا أَن تَزَوَّدا    وَأَن...             0
1  فَما لَبَثٌ يَوماً بِسابِقٍ مَغنَمٍ    وَلا سُ...             0
2  وَإِن تُنظِراني اليَومَ أَقضِ لُبانَةً    وَتَ...             0
3  لَعَمرُكَ ما نَفسٌ بِجِدٍ رَشيدَةٍ    تُؤامِرُ...             0
4  وَإِن ظَهَرَت مِنهُ قَوارِصُ جَمَّةٌ    وَأَفر...             0

In [43]:
#df['البيت'] = df['البيت'].str.replace('.,:','')
#df['البيت'].head()

In [44]:
freq = pd.Series(' '.join(df['البيت']).split()).value_counts()[:10]
print(freq)

في     339840
من     257867
ما     170658
لا     109296
يا      91602
مِن     89260
على     87234
قد      57562
عن      49478
ولا     48257
dtype: int64


In [45]:
freq = list(freq.index)
df['البيت'] = df['البيت'].apply(lambda x: " ".join(x for x in x.split() if x not in freq))
df['البيت'].head()

0    خَليلَيَّ تَستَعجِلا أَن تَزَوَّدا وَأَن تَجمَ...
1    فَما لَبَثٌ يَوماً بِسابِقٍ مَغنَمٍ وَلا سُرعَ...
2    وَإِن تُنظِراني اليَومَ أَقضِ لُبانَةً وَتَستَ...
3    لَعَمرُكَ نَفسٌ بِجِدٍ رَشيدَةٍ تُؤامِرُني سِر...
4    وَإِن ظَهَرَت مِنهُ قَوارِصُ جَمَّةٌ وَأَفرَعَ...
Name: البيت, dtype: object

In [46]:
freq = pd.Series(' '.join(df['البيت']).split()).value_counts()[-10:]
freq

الأثمنُ    1
أحزنُ      1
يمخرُ      1
ترميكمُ    1
فاجأتكم    1
تلهيكمُ    1
تغويكمُ    1
تشديدكم    1
إحجامكم    1
لجنا       1
dtype: int64

In [47]:
df[['البيت','word_count']].head()

البيت  word_count
0  خَليلَيَّ تَستَعجِلا أَن تَزَوَّدا وَأَن تَجمَ...          13
1  فَما لَبَثٌ يَوماً بِسابِقٍ مَغنَمٍ وَلا سُرعَ...          13
2  وَإِن تُنظِراني اليَومَ أَقضِ لُبانَةً وَتَستَ...          12
3  لَعَمرُكَ نَفسٌ بِجِدٍ رَشيدَةٍ تُؤامِرُني سِر...          12
4  وَإِن ظَهَرَت مِنهُ قَوارِصُ جَمَّةٌ وَأَفرَعَ...          13

In [48]:
#df['العصر'].unique()
#df['القافية'].unique()
df['البحر'].unique()

array(['الطويل', 'المنسرح', 'المتقارب', 'الخفيف', 'الكامل', 'السريع',
       'الوافر', 'البسيط', 'الرجز', 'الرمل', 'المجتث', 'المديد', 'الهزج',
       'المتدارك', 'المقتضب', 'المضارع', 'شعر التفعيلة', 'الدوبيت',
       'موشح', 'السلسلة', 'المواليا', 'شعر حر', 'عامي'], dtype=object)

In [49]:
# count values in name column
print(df['العصر'].value_counts()['قبل الإسلام'])
print(df['العصر'].value_counts()['المخضرمين'])
print(df['العصر'].value_counts()['العباسي'])
print(df['العصر'].value_counts()['الأموي'])
print(df['العصر'].value_counts()['المملوكي'])
print(df['العصر'].value_counts()['المغرب والأندلس'])
print(df['العصر'].value_counts()['بين الدولتين'])
print(df['العصر'].value_counts()['الفاطمي'])
print(df['العصر'].value_counts()['الأيوبي'])
print(df['العصر'].value_counts()['الحديث'])
print(df['العصر'].value_counts()['الإسلامي'])
print(df['العصر'].value_counts()['العثماني'])
print ('__________________________________________________')
print(df['البحر'].value_counts()['الطويل'])
print(df['البحر'].value_counts()['المنسرح'])
print(df['البحر'].value_counts()['المتقارب'])
print(df['البحر'].value_counts()['الخفيف'])
print(df['البحر'].value_counts()['الكامل'])
print(df['البحر'].value_counts()['السريع'])
print(df['البحر'].value_counts()['الوافر'])
print(df['البحر'].value_counts()['البسيط'])
print(df['البحر'].value_counts()['الرجز'])
print(df['البحر'].value_counts()['الرمل'])
print(df['البحر'].value_counts()['المجتث'])
print(df['البحر'].value_counts()['المديد'])
print(df['البحر'].value_counts()['الهزج'])
print(df['البحر'].value_counts()['المتدارك'])
print(df['البحر'].value_counts()['المقتضب'])
print(df['البحر'].value_counts()['المضارع'])
print(df['البحر'].value_counts()['شعر التفعيلة'])
print(df['البحر'].value_counts()['الدوبيت'])
print(df['البحر'].value_counts()['موشح'])
print(df['البحر'].value_counts()['السلسلة'])
print(df['البحر'].value_counts()['المواليا'])
print(df['البحر'].value_counts()['شعر حر'])
print(df['البحر'].value_counts()['المواليا'])
print(df['البحر'].value_counts()['عامي'])

21324
29891
231513
61917
158138
107088
18351
122789
110687
669688
2136
151146
__________________________________________________
398215
28052
62463
152561
358813
57081
139310
236636
103101
73261
17591
7650
7306
4507
795
288
164
2710
25356
724
1251
32
1251
6801


In [50]:
# count values in name column
print(df['العصر'].value_counts(['قبل الإسلام'], ['العثماني']))
print ('__________________________________________________')
print(df['القافية'].value_counts(['د'], ['طن']))
print ('__________________________________________________')
print(df['البحر'].value_counts(['الطويل'], ['عامي']))

الحديث             0.397519
العباسي            0.137424
المملوكي           0.093869
العثماني           0.089719
الفاطمي            0.072886
الأيوبي            0.065703
المغرب والأندلس    0.063566
الأموي             0.036753
المخضرمين          0.017743
قبل الإسلام        0.012658
بين الدولتين       0.010893
الإسلامي           0.001268
Name: العصر, dtype: float64
__________________________________________________
ر     0.156803
ل     0.117252
م     0.113198
د     0.111903
ب     0.098296
ن     0.094375
ع     0.039814
ق     0.039118
ء     0.028954
ت     0.027820
ف     0.025489
ح     0.024560
ه     0.022140
س     0.020685
ي     0.017796
ك     0.015719
ج     0.009250
ض     0.007234
ا     0.004886
ط     0.004620
ز     0.003130
ش     0.002928
ص     0.002897
و     0.002748
ث     0.002374
ذ     0.001622
خ     0.001359
غ     0.001234
ظ     0.000959
ى     0.000756
هـ    0.000066
ؤ     0.000005
هن    0.000004
طن    0.000004
لا    0.000002
Name: القافية, dtype: float64
______________________________

In [51]:
# count all values in name column in ascending order
print(df['العصر'].value_counts(ascending=False))

الحديث             669688
العباسي            231513
المملوكي           158138
العثماني           151146
الفاطمي            122789
الأيوبي            110687
المغرب والأندلس    107088
الأموي              61917
المخضرمين           29891
قبل الإسلام         21324
بين الدولتين        18351
الإسلامي             2136
Name: العصر, dtype: int64


In [52]:
df.groupby('العصر')['البحر'].value_counts().unstack().fillna(0)

البحر             البسيط   الخفيف  الدوبيت    الرجز    الرمل   السريع  \
العصر                                                                   
الأموي            7882.0   2806.0      0.0    691.0    835.0    587.0   
الأيوبي          17003.0   6329.0    342.0   5900.0   4156.0   5492.0   
الإسلامي           204.0     56.0      0.0     63.0    119.0      2.0   
الحديث           96423.0  76977.0    179.0  41401.0  40640.0  16969.0   
العباسي          26906.0  23482.0      6.0  17376.0   8050.0  11886.0   
العثماني         23287.0  15092.0    369.0   8122.0   7170.0   4445.0   
الفاطمي          16666.0   9152.0     92.0   8299.0   2258.0   5852.0   
المخضرمين         3777.0    939.0      0.0    344.0    562.0    571.0   
المغرب والأندلس  14478.0   4338.0     60.0   2119.0   3683.0   3828.0   
المملوكي         25478.0  11061.0   1662.0  14960.0   4375.0   5984.0   
بين الدولتين      2026.0   1397.0      0.0   3509.0    673.0    883.0   
قبل الإسلام       2506.0    932.0      0.0    317.0    740.0    582.0   

البحر            السلسلة    الطويل    الكامل  المتدارك  ...  المضارع  المقتضب  \
العصر                                                   ...                     
الأموي               0.0   29334.0    8116.0       0.0  ...      0.0      0.0   
الأيوبي              0.0   28819.0   25782.0     200.0  ...     20.0     22.0   
الإسلامي             0.0    1083.0     234.0       0.0  ...      0.0      0.0   
الحديث             525.0  128709.0  156431.0    2912.0  ...      9.0    488.0   
العباسي              0.0   48843.0   42320.0       7.0  ...     17.0     27.0   
العثماني           161.0   36390.0   31637.0     245.0  ...      8.0      6.0   
الفاطمي              0.0   30331.0   27048.0     144.0  ...     27.0     20.0   
المخضرمين            0.0   12582.0    4164.0       0.0  ...      0.0      0.0   
المغرب والأندلس      0.0   32397.0   24891.0     568.0  ...     13.0    173.0   
المملوكي            38.0   37439.0   32622.0     431.0  ...    194.0     59.0   
بين الدولتين         0.0    4341.0    2379.0       0.0  ...      0.0      0.0   
قبل الإسلام          0.0    7947.0    3189.0       0.0  ...      0.0      0.0   

البحر            المنسرح  المواليا   الهزج   الوافر  شعر التفعيلة  شعر حر  \
العصر                                                                       
الأموي            1122.0       0.0   132.0   8248.0           0.0     0.0   
الأيوبي           2457.0      16.0   552.0   7361.0           0.0     0.0   
الإسلامي            18.0       0.0     0.0    288.0           0.0     0.0   
الحديث            3247.0     292.0  1492.0  55346.0          42.0    20.0   
العباسي          11368.0       0.0  3087.0  21479.0           2.0     0.0   
العثماني          1881.0     565.0   207.0   9970.0          20.0    12.0   
الفاطمي           2817.0      11.0   587.0  10596.0           0.0     0.0   
المخضرمين          395.0       0.0    52.0   4258.0           0.0     0.0   
المغرب والأندلس   1272.0       0.0    62.0   6326.0           0.0     0.0   
المملوكي          2439.0     367.0   584.0  10499.0         100.0     0.0   
بين الدولتين       642.0       0.0   411.0   1396.0           0.0     0.0   
قبل الإسلام        394.0       0.0   140.0   3543.0           0.0     0.0   

البحر              عامي    موشح  
العصر                            
الأموي              0.0     0.0  
الأيوبي             0.0   624.0  
الإسلامي            0.0     0.0  
الحديث           5965.0  9081.0  
العباسي             0.0     8.0  
العثماني          394.0  5479.0  
الفاطمي             1.0    62.0  
المخضرمين           0.0     0.0  
المغرب والأندلس     0.0  6729.0  
المملوكي          441.0  3373.0  
بين الدولتين        0.0     0.0  
قبل الإسلام         0.0     0.0  

[12 rows x 23 columns]

In [53]:
df.groupby('العصر')['القافية'].value_counts().unstack()

القافية                ء    ؤ       ا        ب        ت      ث       ج  \
العصر                                                                    
الأموي             849.0  NaN   149.0   7615.0    909.0   95.0   601.0   
الأيوبي           3016.0  NaN   284.0  10396.0   3489.0  276.0   741.0   
الإسلامي            22.0  NaN     4.0    314.0     42.0    NaN    30.0   
الحديث           24746.0  NaN  2920.0  62577.0  23349.0  948.0  4429.0   
العباسي           4423.0  NaN  1431.0  25709.0   4565.0  745.0  2207.0   
العثماني          4952.0  NaN   855.0  12697.0   3748.0  455.0  1923.0   
الفاطمي           3260.0  8.0   604.0  12043.0   2636.0  393.0  1654.0   
المخضرمين          336.0  NaN   111.0   3256.0    377.0   45.0   250.0   
المغرب والأندلس   3277.0  NaN   514.0  10513.0   1757.0  470.0  1395.0   
المملوكي          3120.0  NaN  1224.0  14586.0   4978.0  442.0  1885.0   
بين الدولتين       467.0  NaN    48.0   3361.0    591.0  117.0   284.0   
قبل الإسلام        310.0  NaN    88.0   2529.0    426.0   14.0   185.0   

القافية                ح      خ        د  ...        ل   لا        م        ن  \
العصر                                     ...                                   
الأموي            1549.0   19.0   5778.0  ...   9814.0  NaN   7634.0   4232.0   
الأيوبي           3512.0  176.0  12442.0  ...  13877.0  NaN  13184.0   9331.0   
الإسلامي            47.0    3.0    110.0  ...    238.0  NaN    164.0    297.0   
الحديث           13729.0  557.0  78503.0  ...  70990.0  NaN  78311.0  70696.0   
العباسي           5408.0  291.0  25135.0  ...  27320.0  NaN  22670.0  20499.0   
العثماني          4047.0  401.0  17322.0  ...  16117.0  NaN  18935.0  13939.0   
الفاطمي           2685.0  175.0  13481.0  ...  16813.0  3.0  12862.0  11625.0   
المخضرمين          741.0    1.0   3110.0  ...   5074.0  NaN   3634.0   1821.0   
المغرب والأندلس   4263.0  276.0  11208.0  ...  12411.0  NaN  11586.0   9379.0   
المملوكي          4082.0  384.0  16196.0  ...  20263.0  NaN  17731.0  14700.0   
بين الدولتين       699.0    NaN   3243.0  ...   1393.0  NaN   1531.0   1038.0   
قبل الإسلام        613.0    7.0   1992.0  ...   3220.0  NaN   2459.0   1434.0   

القافية                ه     هـ   هن       و      ى        ي  
العصر                                                         
الأموي             514.0    NaN  NaN    11.0    8.0   1261.0  
الأيوبي           2449.0    NaN  NaN   316.0  232.0   2539.0  
الإسلامي            11.0    NaN  NaN    10.0    4.0     77.0  
الحديث           19694.0  112.0  6.0  1957.0  376.0  13654.0  
العباسي           3773.0    NaN  NaN   373.0  197.0   2792.0  
العثماني          3295.0    NaN  NaN   693.0   77.0   2348.0  
الفاطمي           2126.0    NaN  NaN   133.0  259.0   1516.0  
المخضرمين          233.0    NaN  NaN    16.0   10.0    663.0  
المغرب والأندلس   1978.0    NaN  NaN   505.0   16.0   1854.0  
المملوكي          2862.0    NaN  NaN   602.0   95.0   2735.0  
بين الدولتين       139.0    NaN  NaN     2.0    NaN    210.0  
قبل الإسلام        225.0    NaN  NaN    11.0    NaN    332.0  

[12 rows x 35 columns]

In [54]:
df.groupby('القافية')['البحر'].value_counts().unstack()

البحر     البسيط   الخفيف  الدوبيت    الرجز   الرمل  السريع  السلسلة   الطويل  \
القافية                                                                         
ء         3691.0  13540.0    118.0   1686.0  1967.0   417.0      NaN   3435.0   
ؤ            NaN      8.0      NaN      NaN     NaN     NaN      NaN      NaN   
ا          264.0    562.0     10.0   2376.0   486.0   281.0      NaN   1234.0   
ب        24942.0  10993.0    185.0   9070.0  6712.0  5404.0     30.0  47950.0   
ت         6814.0   4450.0     78.0   1947.0  2175.0  1450.0      NaN  14516.0   
ث          474.0    250.0     32.0     87.0    63.0   108.0      NaN   1635.0   
ج         2853.0    706.0     30.0    906.0   691.0   532.0      NaN   3092.0   
ح         4137.0   3340.0    100.0   1743.0  2640.0  2463.0     23.0   9272.0   
خ          134.0    211.0     18.0     58.0    26.0   158.0      NaN    975.0   
د        20833.0  16208.0    198.0  14566.0  7329.0  5189.0     77.0  49869.0   
ذ          207.0    182.0     24.0    292.0   151.0    82.0      NaN    481.0   
ر        38004.0  20176.0    262.0  15024.0  9140.0  9621.0     20.0  70497.0   
ز          433.0   1205.0     22.0    605.0    80.0   128.0      NaN    911.0   
س         4668.0   3055.0     46.0   2018.0  2126.0  2007.0      NaN   6823.0   
ش          459.0    531.0     32.0    289.0   409.0   225.0      NaN   1119.0   
ص          297.0    628.0     22.0    238.0   195.0   180.0      NaN   1373.0   
ض         1202.0   1689.0     49.0    901.0   484.0   574.0      NaN   3137.0   
ط         1000.0    521.0     28.0    685.0   392.0   262.0      NaN   2960.0   
طن           NaN      NaN      NaN      NaN     6.0     NaN      NaN      NaN   
ظ          196.0    178.0     20.0     61.0    21.0    44.0     28.0    381.0   
ع         6498.0   3566.0     70.0   3556.0  2399.0  2267.0      NaN  22943.0   
غ          213.0    158.0     12.0    110.0   112.0   146.0      NaN    759.0   
ف         7047.0   3206.0     91.0   2347.0  1448.0  1096.0      NaN  10694.0   
ق         7587.0   5989.0    161.0   6838.0  2892.0  2801.0      NaN  14469.0   
ك         3849.0   2543.0     99.0   2198.0  1420.0  1124.0     41.0   4356.0   
ل        27708.0  18504.0    265.0  11716.0  8060.0  6135.0    403.0  46242.0   
لا           NaN      NaN      NaN      NaN     NaN     NaN      NaN      NaN   
م        31319.0  15131.0    213.0   8860.0  8483.0  5841.0     14.0  46469.0   
ن        28775.0  16878.0    207.0   9443.0  9079.0  5468.0     77.0  18109.0   
ه         9054.0   4734.0     98.0   3080.0  2246.0  1857.0     11.0   3604.0   
هـ         112.0      NaN      NaN      NaN     NaN     NaN      NaN      NaN   
هن           NaN      NaN      NaN      NaN     NaN     NaN      NaN      NaN   
و          163.0    292.0     14.0    349.0   150.0   157.0      NaN   2110.0   
ى           26.0     10.0      NaN    320.0   102.0    52.0      NaN    344.0   
ي         3677.0   3117.0    206.0   1732.0  1777.0  1012.0      NaN   8456.0   

البحر     الكامل  المتدارك  ...  المضارع  المقتضب  المنسرح  المواليا   الهزج  \
القافية                     ...                                                
ء        16966.0       NaN  ...      NaN      NaN    251.0       NaN    47.0   
ؤ            NaN       NaN  ...      NaN      NaN      NaN       NaN     NaN   
ا         1265.0      11.0  ...      NaN      2.0     62.0       NaN    24.0   
ب        28092.0     278.0  ...      5.0    280.0   4077.0      42.0   611.0   
ت         8890.0      30.0  ...     10.0      4.0    406.0      42.0   247.0   
ث          633.0       NaN  ...      NaN      NaN     50.0       NaN     NaN   
ج         3810.0     444.0  ...      NaN      7.0    267.0      11.0   144.0   
ح         9483.0      76.0  ...      NaN      2.0    424.0      34.0   137.0   
خ          412.0       NaN  ...      NaN      NaN     20.0       NaN    10.0   
د        43656.0    1064.0  ...     27.0     12.0   3017.0     128.0   533.0   
ذ         1028.0      11.0  ...      NaN  

In [55]:
df['word_count'] = df['البيت'].apply(lambda x: len(str(x).split(" ")))
df[['البيت','word_count']].head()

البيت  word_count
0  خَليلَيَّ تَستَعجِلا أَن تَزَوَّدا وَأَن تَجمَ...           9
1  فَما لَبَثٌ يَوماً بِسابِقٍ مَغنَمٍ وَلا سُرعَ...          10
2  وَإِن تُنظِراني اليَومَ أَقضِ لُبانَةً وَتَستَ...           9
3  لَعَمرُكَ نَفسٌ بِجِدٍ رَشيدَةٍ تُؤامِرُني سِر...           8
4  وَإِن ظَهَرَت مِنهُ قَوارِصُ جَمَّةٌ وَأَفرَعَ...           9

In [56]:
df.groupby('العصر')['word_count'].value_counts().unstack()

word_count         2      3       4        5        6        7         8   \
العصر                                                                       
الأموي            NaN    NaN    48.0    189.0   1116.0   3760.0   12834.0   
الأيوبي          16.0   31.0   614.0   1861.0   6425.0  12082.0   26841.0   
الإسلامي          NaN    NaN     2.0      4.0     37.0    166.0     531.0   
الحديث           38.0  110.0  4100.0  12330.0  42998.0  77712.0  175963.0   
العباسي           1.0    5.0   987.0   3753.0  16958.0  32321.0   60683.0   
العثماني          8.0   25.0   988.0   3124.0   9972.0  16585.0   36512.0   
الفاطمي           NaN    9.0   589.0   1370.0   6835.0  13652.0   31117.0   
المخضرمين         NaN    NaN    10.0     72.0    619.0   2369.0    7007.0   
المغرب والأندلس   3.0   70.0   985.0   2466.0   6445.0  10350.0   24628.0   
المملوكي          5.0   68.0  1023.0   2674.0   8819.0  16159.0   38498.0   
بين الدولتين      NaN    NaN    40.0    234.0    831.0   1943.0    4815.0   
قبل الإسلام       NaN    NaN     5.0     74.0    522.0   1856.0    5566.0   

word_count             9         10       11       12      13     14    15  \
العصر                                                                        
الأموي            18981.0   15937.0   6963.0   1802.0   264.0   20.0   3.0   
الأيوبي           29849.0   22292.0   8121.0   2231.0   272.0   46.0   4.0   
الإسلامي            664.0     495.0    189.0     41.0     6.0    1.0   NaN   
الحديث           182364.0  119826.0  42630.0  10090.0  1314.0  193.0  16.0   
العباسي           59138.0   39255.0  14129.0   3672.0   535.0   66.0   5.0   
العثماني          40267.0   29067.0  11213.0   2863.0   456.0   56.0   7.0   
الفاطمي           33864.0   23825.0   8873.0   2291.0   324.0   37.0   2.0   
المخضرمين          9141.0    7057.0   2848.0    678.0    80.0    9.0   1.0   
المغرب والأندلس   28424.0   22112.0   8695.0   2480.0   382.0   46.0   1.0   
المملوكي          43413.0   31569.0  11989.0   3347.0   472.0   87.0   8.0   
بين الدولتين       5273.0    3491.0   1358.0    316.0    45.0    5.0   NaN   
قبل الإسلام        6532.0    4538.0   1769.0    398.0    55.0    5.0   3.0   

word_count        16   17   22  
العصر                           
الأموي           NaN  NaN  NaN  
الأيوبي          2.0  NaN  NaN  
الإسلامي         NaN  NaN  NaN  
الحديث           4.0  NaN  NaN  
العباسي          4.0  NaN  1.0  
العثماني         1.0  2.0  NaN  
الفاطمي          1.0  NaN  NaN  
المخضرمين        NaN  NaN  NaN  
المغرب والأندلس  1.0  NaN  NaN  
المملوكي         7.0  NaN  NaN  
بين الدولتين     NaN  NaN  NaN  
قبل الإسلام      1.0  NaN  NaN

In [57]:
df.groupby('البحر')['word_count'].value_counts().unstack()

word_count      2      3       4       5        6        7         8   \
البحر                                                                   
البسيط         1.0   12.0   158.0   295.0   3342.0   8893.0   36052.0   
الخفيف         NaN    7.0   448.0  1395.0   5432.0  20541.0   50717.0   
الدوبيت        NaN    NaN     2.0     7.0     99.0    220.0     682.0   
الرجز         13.0   20.0  2689.0  5693.0  14453.0  21017.0   30480.0   
الرمل          3.0   11.0   934.0  3938.0  11513.0  14224.0   23114.0   
السريع         2.0    NaN    75.0   233.0   5324.0  15543.0   20548.0   
السلسلة        NaN    NaN     NaN     8.0      5.0      7.0     115.0   
الطويل         NaN    NaN    38.0    64.0    413.0   4476.0   50424.0   
الكامل         4.0   24.0   479.0  4525.0  23952.0  46538.0  112033.0   
المتدارك       NaN    9.0   225.0   169.0   1315.0   1559.0     986.0   
المتقارب       NaN    3.0   116.0   242.0   5151.0  14865.0   24827.0   
المجتث        15.0   10.0   843.0  4078.0   9201.0   2897.0     457.0   
المديد         NaN    NaN    32.0   274.0   1905.0   2547.0    1952.0   
المضارع        NaN    NaN    10.0    24.0     58.0     56.0      70.0   
المقتضب        NaN    NaN   145.0   245.0    257.0     86.0      28.0   
المنسرح        NaN   12.0   335.0   361.0   1603.0   5540.0   10319.0   
المواليا       NaN    NaN    31.0    32.0     33.0     25.0     106.0   
الهزج          NaN    NaN   246.0   954.0   3151.0   2019.0     784.0   
الوافر         NaN    3.0   194.0  1044.0   7110.0  22761.0   56245.0   
شعر التفعيلة   NaN    NaN     4.0    44.0     68.0     24.0       4.0   
شعر حر         NaN    NaN     3.0    10.0      3.0      NaN       NaN   
عامي           5.0    4.0   117.0   282.0    622.0    666.0    1372.0   
موشح          28.0  203.0  2267.0  4234.0   6567.0   4451.0    3680.0   

word_count          9         10       11       12      13     14    15    16  \
البحر                                                                           
البسيط         71047.0   74106.0  32774.0   8559.0  1209.0  169.0  10.0   8.0   
الخفيف         45942.0   21994.0   5310.0    710.0    59.0    6.0   NaN   NaN   
الدوبيت          880.0     577.0    200.0     39.0     3.0    1.0   NaN   NaN   
الرجز          19378.0    7376.0   1631.0    312.0    32.0    5.0   1.0   1.0   
الرمل          14053.0    4511.0    796.0    143.0    18.0    3.0   NaN   NaN   
السريع         11431.0    3384.0    489.0     47.0     5.0    NaN   NaN   NaN   
السلسلة          214.0     230.0    108.0     26.0     9.0    2.0   NaN   NaN   
الطويل        122066.0  135575.0  64085.0  18147.0  2569.0  321.0  26.0  10.0   
الكامل        108402.0   50822.0  10392.0   1469.0   150.0   20.0   2.0   NaN   
المتدارك         205.0      33.0      4.0      2.0     NaN    NaN   NaN   NaN   
المتقارب       12869.0    3825.0    492.0     67.0     3.0    2.0   1.0   NaN   
المجتث            58.0      17.0      8.0      6.0     1.0    NaN   NaN   NaN   
المديد           753.0     158.0     27.0      2.0     NaN    NaN   NaN   NaN   
المضارع           44.0      17.0      8.0      1.0     NaN    NaN   NaN   NaN   
المقتضب           19.0      10.0      3.0      2.0     NaN    NaN   NaN   NaN   
المنسرح         6921.0    2471.0    434.0     52.0     3.0    1.0   NaN   NaN   
المواليا         187.0     316.0    302.0    173.0    31.0   10.0   3.0   2.0   
الهزج            102.0      32.0     11.0      4.0     2.0    1.0   NaN   NaN   
الوافر         39540.0   11575.0    749.0     67.0    12.0    7.0   3.0   NaN   
شعر التفعيلة       1.0       6.0      5.0      6.0     2.0    NaN   NaN   NaN   
شعر حر             2.0      10.0      3.0      1.0     NaN    NaN   NaN   NaN   
عامي            1829.0    1338.0    440.0    104.0    15.0    4.0   3.0   NaN   
موشح            1967.0    1081.0    506.0    270.0    82.0   19.0   1.0   NaN   

word_count     17   22  
البحر                   
البسيط        1.0  NaN  
الخفيف        NaN  NaN  
الدوبيت       NaN  NaN  
الرجز         NaN  NaN 

# TF-IDF and BoW (TextBlob)

In [58]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1684668 entries, 0 to 1684667
Data columns (total 14 columns):
 #   Column        Non-Null Count    Dtype  
---  ------        --------------    -----  
 0   Unnamed: 0.1  1684668 non-null  int64  
 1   Unnamed: 0    1684668 non-null  int64  
 2   العصر         1684668 non-null  object 
 3   الشاعر        1684668 non-null  object 
 4   الديوان       1684668 non-null  object 
 5   القافية       1684668 non-null  object 
 6   البحر         1684668 non-null  object 
 7   الشطر الايسر  1684668 non-null  object 
 8   الشطر الايمن  1684668 non-null  object 
 9   البيت         1684668 non-null  object 
 10  word_count    1684668 non-null  int64  
 11  avg_word      1684668 non-null  float64
 12  stopwords     1684668 non-null  int64  
 13  punctuations  1684668 non-null  int64  
dtypes: float64(1), int64(5), object(8)
memory usage: 179.9+ MB


In [59]:
%pip install numpy

Note: you may need to restart the kernel to use updated packages.


In [60]:
%pip install scikit-learn

Note: you may need to restart the kernel to use updated packages.


In [61]:
%pip install pandas

Note: you may need to restart the kernel to use updated packages.


In [62]:
import numpy as np
import pandas as pd
df = pd.read_csv("C:\\Users\\snmuj\\Documents\\NLP--Course1\\clean_preislamic-poems.csv")
df.head()

Unnamed: 0.2  Unnamed: 0.1  Unnamed: 0        العصر             الشاعر  \
0             0             0           0  قبل الإسلام  عمرو بنِ قُمَيئَة   
1             1             1           1  قبل الإسلام  عمرو بنِ قُمَيئَة   
2             2             2           2  قبل الإسلام  عمرو بنِ قُمَيئَة   
3             3             3           3  قبل الإسلام  عمرو بنِ قُمَيئَة   
4             4             4           4  قبل الإسلام  عمرو بنِ قُمَيئَة   

           الديوان القافية   البحر                           الشطر الايسر  \
0  الديوان الرئيسي       د  الطويل   وَأَن تَجمَعا شَملي وَتَنتَظِرا غَدا   
1  الديوان الرئيسي       د  الطويل  وَلا سُرعَتي يَوماً بِسابِقَةِ الرَدى   
2  الديوان الرئيسي       د  الطويل  وَتَستَوجِبا مَنّاً عَلَيَّ وَتُحمَدا   
3  الديوان الرئيسي       د  الطويل    تُؤامِرُني سِرّاً لِأَصرِمَ مَرثَدا   
4  الديوان الرئيسي       د  الطويل   وَأَفرَعَ في لَومي مِراراً وَأَصعَدا   

                             الشطر الايمن  \
0   خَليلَيَّ لا تَستَعجِلا أَن تَزَوَّدا   
1     فَما لَبَثٌ يَوماً بِسابِقٍ مَغنَمٍ   
2  وَإِن تُنظِراني اليَومَ أَقضِ لُبانَةً   
3      لَعَمرُكَ ما نَفسٌ بِجِدٍ رَشيدَةٍ   
4    وَإِن ظَهَرَت مِنهُ قَوارِصُ جَمَّةٌ   

                                               البيت  
0  خَليلَيَّ لا تَستَعجِلا أَن تَزَوَّدا    وَأَن...  
1  فَما لَبَثٌ يَوماً بِسابِقٍ مَغنَمٍ    وَلا سُ...  
2  وَإِن تُنظِراني اليَومَ أَقضِ لُبانَةً    وَتَ...  
3  لَعَمرُكَ ما نَفسٌ بِجِدٍ رَشيدَةٍ    تُؤامِرُ...  
4  وَإِن ظَهَرَت مِنهُ قَوارِصُ جَمَّةٌ    وَأَفر...

In [63]:
df['البيت']

0        خَليلَيَّ لا تَستَعجِلا أَن تَزَوَّدا    وَأَن...
1        فَما لَبَثٌ يَوماً بِسابِقٍ مَغنَمٍ    وَلا سُ...
2        وَإِن تُنظِراني اليَومَ أَقضِ لُبانَةً    وَتَ...
3        لَعَمرُكَ ما نَفسٌ بِجِدٍ رَشيدَةٍ    تُؤامِرُ...
4        وَإِن ظَهَرَت مِنهُ قَوارِصُ جَمَّةٌ    وَأَفر...
                               ...                        
21319    البيضُ حصنٌ لهم إذا فزِعُوا    وسابغاتٌ كأنَّه...
21320    والبيضُ قد ثُلِّمَت مضارِبُها    بها نُفوسُ ال...
21321    عليك بأنقاض العراق فقد علت    عليك بتخدين النس...
21322    لعمري لقد راش ابن سعدة نفسه    بريش الدنابي لا...
21323    بنى لك معروف بناء هدمته    وللشرف العاديّ بان ...
Name: البيت, Length: 21324, dtype: object

# Vectorization 

In [64]:
from sklearn.feature_extraction.text import CountVectorizer
# create the transform
vectorizer = CountVectorizer(analyzer = "word")
# tokenize and build vocab
vectorizer.fit(df['البيت'])
# summarize
print(vectorizer.vocabulary_)
# encode document
vector = vectorizer.transform(df['البيت'])
# summerize encoded vector
print(vector.shape)
print(type(vector))
print(vector.toarray())

{'ليل': 15208, 'لا': 14378, 'ست': 9570, 'عج': 11504, 'دا': 7866, 'جم': 6729, 'عا': 11310, 'ملي': 16075, 'نت': 16585, 'را': 8479, 'ما': 15265, 'وما': 19313, 'ساب': 9426, 'غن': 12295, 'رع': 8874, 'تي': 6172, 'الر': 2515, 'دى': 8171, 'نظ': 16852, 'راني': 8622, 'الي': 4069, 'وم': 19310, 'قض': 13592, 'بان': 4593, 'وج': 18555, 'با': 4273, 'حم': 7318, 'مر': 15607, 'فس': 12901, 'شيد': 10394, 'ؤام': 1213, 'ني': 17090, 'صر': 10572, 'رث': 8701, 'نه': 16992, 'وار': 17842, 'فر': 12833, 'في': 13210, 'ومي': 19414, 'رارا': 8542, 'صع': 10603, 'لى': 15155, 'ير': 19963, 'نب': 16551, 'كون': 14269, 'يت': 19775, 'وى': 19531, 'ول': 19214, 'باغ': 4358, 'كاد': 13861, 'مري': 15675, 'عم': 11872, 'الم': 3474, 'رء': 8431, 'دعو': 8032, 'بل': 5069, 'ذا': 8229, 'نادي': 16434, 'قام': 13374, 'ظيم': 11308, 'ماد': 15296, 'الق': 3253, 'در': 7993, 'ؤي': 1262, 'نها': 16993, 'وق': 19111, 'حل': 7273, 'الريح': 2619, 'تر': 5615, 'ذي': 8409, 'المال': 3501, 'رف': 8897, 'رت': 8685, 'طء': 10955, 'والي': 18335, 'طم': 11139, 'ذو': 83

(21324, 20565)
<class 'scipy.sparse.csr.csr_matrix'>
[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]


# Train, Test and Predict X(verses), y(metrics)

In [65]:
from sklearn import preprocessing
le = preprocessing.LabelEncoder()
df_le = le.fit_transform(df['البحر'])
df_le

array([5, 5, 5, ..., 5, 5, 5])

In [66]:
from sklearn.feature_extraction import DictVectorizer
df = pd.read_csv("C:\\Users\\snmuj\\Documents\\NLP--Course1\\clean_preislamic-poems.csv")
dv = DictVectorizer(sparse = False)
print(df['البحر'].value_counts().to_dict())
dv.fit_transform(df['البحر'].value_counts().to_dict())

{'الطويل': 7947, 'الوافر': 3543, 'الكامل': 3189, 'البسيط': 2506, 'الخفيف': 932, 'المتقارب': 892, 'الرمل': 740, 'السريع': 582, 'المنسرح': 394, 'الرجز': 317, 'المديد': 141, 'الهزج': 140, 'المجتث': 1}


array([[2.506e+03, 9.320e+02, 3.170e+02, 7.400e+02, 5.820e+02, 7.947e+03,
        3.189e+03, 8.920e+02, 1.000e+00, 1.410e+02, 3.940e+02, 1.400e+02,
        3.543e+03]])

In [67]:
from sklearn.model_selection import train_test_split
X = df['البيت']
y = df['البحر']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=0)

In [68]:
from sklearn.feature_extraction.text import CountVectorizer
count_vect = CountVectorizer()
X_train_counts = count_vect.fit_transform(X_train)
X_train_counts.shape

(14287, 16145)

## Train a Classifier
LinearSVC handles sparse input better, and scales well to large numbers of samples.

In [69]:
from sklearn.svm import LinearSVC
clf = LinearSVC()
clf.fit(X_train_counts,y_train)

LinearSVC()

In [70]:
X

0        خَليلَيَّ لا تَستَعجِلا أَن تَزَوَّدا    وَأَن...
1        فَما لَبَثٌ يَوماً بِسابِقٍ مَغنَمٍ    وَلا سُ...
2        وَإِن تُنظِراني اليَومَ أَقضِ لُبانَةً    وَتَ...
3        لَعَمرُكَ ما نَفسٌ بِجِدٍ رَشيدَةٍ    تُؤامِرُ...
4        وَإِن ظَهَرَت مِنهُ قَوارِصُ جَمَّةٌ    وَأَفر...
                               ...                        
21319    البيضُ حصنٌ لهم إذا فزِعُوا    وسابغاتٌ كأنَّه...
21320    والبيضُ قد ثُلِّمَت مضارِبُها    بها نُفوسُ ال...
21321    عليك بأنقاض العراق فقد علت    عليك بتخدين النس...
21322    لعمري لقد راش ابن سعدة نفسه    بريش الدنابي لا...
21323    بنى لك معروف بناء هدمته    وللشرف العاديّ بان ...
Name: البيت, Length: 21324, dtype: object

## Build a Pipeline
Remember that only our training set has been vectorized into a full vocabulary. In order to perform an analysis on our test set we'll have to submit it to the same procedures. Fortunately scikit-learn offers a [**Pipeline**](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html) class that behaves like a compound classifier.

In [71]:
# 'fit' is used for training not for testing. For testing we use 'predict'
from sklearn.preprocessing import StandardScaler
vect = CountVectorizer().fit_transform(X_train)
X_vec= StandardScaler(with_mean=False).fit_transform(vect)
clf = LinearSVC()
clf.fit(X_vec, y_train)

C:\Users\snmuj\anaconda3\lib\site-packages\sklearn\svm\_base.py:1206: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


LinearSVC()

In [72]:
from sklearn.pipeline import Pipeline
# from sklearn.feature_extraction.text import TfidfVectorizer
# from sklearn.feature_extraction.text import CountVectorizer
# from sklearn.svm import LinearSVC
text_clf = Pipeline([('vect', CountVectorizer()),
                     ('clf', LinearSVC()),
])
# Feed the training data through the pipeline
text_clf.fit(X_train, y_train)  

Pipeline(steps=[('vect', CountVectorizer()), ('clf', LinearSVC())])

In [73]:
from sklearn.svm import LinearSVC
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.datasets import make_classification
X, y = make_classification(n_features=4, random_state=0)
clf = make_pipeline(StandardScaler(),
                    LinearSVC(random_state=0, tol=1e-5))
clf.fit(X, y)
Pipeline(steps=[('standardscaler', StandardScaler()),
                ('linearsvc', LinearSVC(random_state=0, tol=1e-05))])

Pipeline(steps=[('standardscaler', StandardScaler()),
                ('linearsvc', LinearSVC(random_state=0, tol=1e-05))])

## Test the classifier and display results

In [74]:
# Form a prediction set
predictions = text_clf.predict(X_test)

In [75]:
# Report the confusion matrix
from sklearn import metrics
print(metrics.confusion_matrix(y_test,predictions))

[[ 171   23    6   17   16  278  120   29    0    3   13    3  137]
 [  16   56    4    7    7   79   39   13    0    0    3    1   72]
 [   8    6    7    3    3   38   17    6    0    0    2    0   19]
 [  25    8    5   33    3   75   42    6    0    2    5    0   49]
 [  26    3    0   11   24   52   38    9    0    1    4    1   42]
 [ 229   54   18   44   35 1524  339   60    0    3   20    2  338]
 [  56   25    9   32   17  335  316   38    0    7    7    6  185]
 [  18   17    5    6    4   79   45   53    0    1    3    2   48]
 [   0    0    0    0    0    1    0    0    0    0    0    0    0]
 [   5    0    0    2    1   12   10    4    0    7    1    0    9]
 [  13    7    1    5    3   37   20    3    0    1   14    0   17]
 [   4    0    1    2    1    9    4    2    0    1    1    2   15]
 [  80   38    7   32   21  327  158   41    0    1    7    5  439]]


In [76]:
# Print a classification report
print(metrics.classification_report(y_test,predictions))

              precision    recall  f1-score   support

      البسيط       0.26      0.21      0.23       816
      الخفيف       0.24      0.19      0.21       297
       الرجز       0.11      0.06      0.08       109
       الرمل       0.17      0.13      0.15       253
      السريع       0.18      0.11      0.14       211
      الطويل       0.54      0.57      0.55      2666
      الكامل       0.28      0.31      0.29      1033
    المتقارب       0.20      0.19      0.19       281
      المجتث       0.00      0.00      0.00         1
      المديد       0.26      0.14      0.18        51
     المنسرح       0.17      0.12      0.14       121
       الهزج       0.09      0.05      0.06        42
      الوافر       0.32      0.38      0.35      1156

    accuracy                           0.38      7037
   macro avg       0.22      0.19      0.20      7037
weighted avg       0.36      0.38      0.37      7037



C:\Users\snmuj\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\snmuj\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\snmuj\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [77]:
df

Unnamed: 0.2  Unnamed: 0.1  Unnamed: 0        العصر  \
0                 0             0           0  قبل الإسلام   
1                 1             1           1  قبل الإسلام   
2                 2             2           2  قبل الإسلام   
3                 3             3           3  قبل الإسلام   
4                 4             4           4  قبل الإسلام   
...             ...           ...         ...          ...   
21319       1681060       1826323     1826323  قبل الإسلام   
21320       1681061       1826324     1826324  قبل الإسلام   
21321       1683929       1830596     1830596  قبل الإسلام   
21322       1683930       1830597     1830597  قبل الإسلام   
21323       1683931       1830598     1830598  قبل الإسلام   

                      الشاعر                      الديوان القافية    البحر  \
0          عمرو بنِ قُمَيئَة              الديوان الرئيسي       د   الطويل   
1          عمرو بنِ قُمَيئَة              الديوان الرئيسي       د   الطويل   
2          عمرو بنِ قُمَيئَة              الديوان الرئيسي       د   الطويل   
3          عمرو بنِ قُمَيئَة              الديوان الرئيسي       د   الطويل   
4          عمرو بنِ قُمَيئَة              الديوان الرئيسي       د   الطويل   
...                      ...                          ...     ...      ...   
21319            درهم بن زيد              الديوان الرئيسي       ف  المنسرح   
21320            درهم بن زيد              الديوان الرئيسي       ف  المنسرح   
21321  سعدة بنت فريد الأسدية  ديوان سعدة بنت فريد الأسدية       م   الطويل   
21322  سعدة بنت فريد الأسدية  ديوان سعدة بنت فريد الأسدية       م   الطويل   
21323  سعدة بنت فريد الأسدية  ديوان سعدة بنت فريد الأسدية       م   الطويل   

                                الشطر الايسر  \
0       وَأَن تَجمَعا شَملي وَتَنتَظِرا غَدا   
1      وَلا سُرعَتي يَوماً بِسابِقَةِ الرَدى   
2      وَتَستَوجِبا مَنّاً عَلَيَّ وَتُحمَدا   
3        تُؤامِرُني سِرّاً لِأَصرِمَ مَرثَدا   
4       وَأَفرَعَ في لَومي مِراراً وَأَصعَدا   
...                                      ...   
21319              وسابغاتٌ كأنَّها النُطَفُ   
21320          بها نُفوسُ الكُماةِ تُختَطَفُ   
21321             عليك بتخدين النساء الكرائم   
21322           بريش الدنابي لا بريش القوادم   
21323               وللشرف العاديّ بان وهادم   

                                 الشطر الايمن  \
0       خَليلَيَّ لا تَستَعجِلا أَن تَزَوَّدا   
1         فَما لَبَثٌ يَوماً بِسابِقٍ مَغنَمٍ   
2      وَإِن تُنظِراني اليَومَ أَقضِ لُبانَةً   
3          لَعَمرُكَ ما نَفسٌ بِجِدٍ رَشيدَةٍ   
4        وَإِن ظَهَرَت مِنهُ قَوارِصُ جَمَّةٌ   
...                                       ...   
21319             البيضُ حصنٌ لهم إذا فزِعُوا   
21320           والبيضُ قد ثُلِّمَت مضارِبُها   
21321              عليك بأنقاض العراق فقد علت   
21322             لعمري لقد راش ابن سعدة نفسه   
21323                 بنى لك معروف بناء هدمته   

                                                   البيت  
0      خَليلَيَّ لا تَستَعجِلا أَن تَزَوَّدا    وَأَن...  
1      فَما لَبَثٌ يَوماً بِسابِقٍ مَغنَمٍ    وَلا سُ...  
2      وَإِن تُنظِراني اليَومَ أَقضِ لُبانَةً    وَتَ...  
3      لَعَمرُكَ ما نَفسٌ بِجِدٍ رَشيدَةٍ    تُؤامِرُ...  
4      وَإِن ظَهَرَت مِنهُ قَوارِصُ جَمَّةٌ    وَأَفر...  
...                                                  ...  
21319  البيضُ حصنٌ لهم إذا فزِعُوا    وسابغاتٌ كأنَّه...  
21320  والبيضُ قد ثُلِّمَت مضارِبُها    بها نُفوسُ ال...  
21321  عليك بأنقاض العراق فقد علت    عليك بتخدين النس...  
21322  لعمري لقد راش ابن سعدة نفسه    بريش الدنابي لا...  
21323  بنى لك معروف بناء هدمته    وللشرف العاديّ بان ...  

[21324 rows x 11 columns]

## Split the data into train & test sets:

In [78]:
from sklearn.model_selection import train_test_split
X = df['البيت']
y = df['البحر']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

## Build pipelines to vectorize the data, then train and fit a model
Now that we have sets to train and test, we'll develop a selection of pipelines, each with a different model.

In [79]:
from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.svm import LinearSVC

# Naïve Bayes:
text_clf_nb = Pipeline([('vect', CountVectorizer()),
                     ('clf', MultinomialNB()),
])

# Linear SVC:
text_clf_lsvc = Pipeline([('vect', CountVectorizer()),
                     ('clf', LinearSVC()),
])

## Feed the training data through the first pipeline
We'll run naïve Bayes first

In [80]:
text_clf_nb.fit(X_train, y_train)

Pipeline(steps=[('vect', CountVectorizer()), ('clf', MultinomialNB())])

## Run predictions and analyze the results (naïve Bayes)

In [81]:
# Form a prediction set
predictions = text_clf_nb.predict(X_test)

In [82]:
# Report the confusion matrix
from sklearn import metrics
print(metrics.confusion_matrix(y_test,predictions))

[[  82    0    0    1    1  722   19    0    0    0    0   32]
 [   7    0    0    0    0  270    7    0    0    0    0   25]
 [   7    0    0    0    0   89    7    0    0    0    0    1]
 [   3    0    0    1    0  208    7    0    0    0    0   17]
 [  11    0    0    0    1  161    9    0    0    0    0    7]
 [  65    1    0    2    0 2459   31    0    0    0    0   66]
 [  43    0    0    1    0  895   76    0    0    0    0   30]
 [  11    0    0    0    0  286    7    0    0    0    0   26]
 [   2    0    0    0    0   38    0    0    0    0    0    1]
 [  11    0    1    0    0  119    3    0    0    1    0    1]
 [   5    0    0    0    0   36    1    0    0    0    0    4]
 [  33    0    0    0    0  934   17    1    0    0    0  135]]


In [83]:
# Print a classification report
print(metrics.classification_report(y_test,predictions))

              precision    recall  f1-score   support

      البسيط       0.29      0.10      0.14       857
      الخفيف       0.00      0.00      0.00       309
       الرجز       0.00      0.00      0.00       104
       الرمل       0.20      0.00      0.01       236
      السريع       0.50      0.01      0.01       189
      الطويل       0.40      0.94      0.56      2624
      الكامل       0.41      0.07      0.12      1045
    المتقارب       0.00      0.00      0.00       330
      المديد       0.00      0.00      0.00        41
     المنسرح       1.00      0.01      0.01       136
       الهزج       0.00      0.00      0.00        46
      الوافر       0.39      0.12      0.18      1120

    accuracy                           0.39      7037
   macro avg       0.27      0.10      0.09      7037
weighted avg       0.35      0.39      0.27      7037



C:\Users\snmuj\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\snmuj\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\snmuj\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [84]:
# Print the overall accuracy
print(metrics.accuracy_score(y_test,predictions))

0.39150206053716075


Naïve Bayes gave us better-than-average results at 76.4% for classifying reviews as positive or negative based on text alone. Let's see if we can do better.

## Feed the training data through the second pipeline
Next we'll run Linear SVC

In [85]:
text_clf_lsvc.fit(X_train, y_train)

Pipeline(steps=[('vect', CountVectorizer()), ('clf', LinearSVC())])

## Run predictions and analyze the results (Linear SVC)

In [86]:
# Form a prediction set
predictions = text_clf_lsvc.predict(X_test)

In [87]:
# Report the confusion matrix
from sklearn import metrics
print(metrics.confusion_matrix(y_test,predictions))

[[ 202   24    5   28   20  309  104   18    5   12    0  130]
 [  29   50    3    9   15   73   43   14    1    3    0   69]
 [   8    1    7    3    0   35   25    5    1    1    0   18]
 [  21    7    3   35    5   77   28    6    3    1    2   48]
 [  21    4    0    8   17   57   45    8    2    3    0   24]
 [ 193   59   16   42   34 1496  328   59    9   32    5  351]
 [  71   21   15   35   18  339  303   28    8    8    2  197]
 [  36   12    2    7    7   86   55   51    2    4    1   67]
 [   3    0    0    2    1   12   11    0    6    2    0    4]
 [  10    6    2    3    4   46   26    2    3   14    0   20]
 [   4    2    0    0    1    7    9    2    0    1    2   18]
 [  73   42    9   34   11  316  146   34    6    5    2  442]]


In [88]:
# Print a classification report
print(metrics.classification_report(y_test,predictions))

              precision    recall  f1-score   support

      البسيط       0.30      0.24      0.26       857
      الخفيف       0.22      0.16      0.19       309
       الرجز       0.11      0.07      0.08       104
       الرمل       0.17      0.15      0.16       236
      السريع       0.13      0.09      0.11       189
      الطويل       0.52      0.57      0.55      2624
      الكامل       0.27      0.29      0.28      1045
    المتقارب       0.22      0.15      0.18       330
      المديد       0.13      0.15      0.14        41
     المنسرح       0.16      0.10      0.13       136
       الهزج       0.14      0.04      0.07        46
      الوافر       0.32      0.39      0.35      1120

    accuracy                           0.37      7037
   macro avg       0.23      0.20      0.21      7037
weighted avg       0.36      0.37      0.36      7037



In [89]:
# Print the overall accuracy
print(metrics.accuracy_score(y_test,predictions))

0.3730282790962058


## Take a quick look at the `label` column:

In [90]:
df['البحر'].value_counts()

الطويل      7947
الوافر      3543
الكامل      3189
البسيط      2506
الخفيف       932
المتقارب     892
الرمل        740
السريع       582
المنسرح      394
الرجز        317
المديد       141
الهزج        140
المجتث         1
Name: البحر, dtype: int64

## Split the data into train & test sets:

In [91]:
from sklearn.model_selection import train_test_split
X = df['البيت']
y = df['البحر']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

## Build pipelines to vectorize the data, then train and fit a model
Now that we have sets to train and test, we'll develop a selection of pipelines, each with a different model.

In [92]:
from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.svm import LinearSVC
# Naïve Bayes:
text_clf_nb = Pipeline([('vect', CountVectorizer()),
                     ('clf', MultinomialNB()),
])
# Linear SVC:
text_clf_lsvc = Pipeline([('vect', CountVectorizer()),
                     ('clf', LinearSVC()),
])

## Feed the training data through the first pipeline
We'll run naïve Bayes first

In [93]:
text_clf_nb.fit(X_train, y_train)

Pipeline(steps=[('vect', CountVectorizer()), ('clf', MultinomialNB())])

## Run predictions and analyze the results (naïve Bayes)

In [94]:
# Form a prediction set
predictions = text_clf_nb.predict(X_test)

In [95]:
# Report the confusion matrix
from sklearn import metrics
print(metrics.confusion_matrix(y_test,predictions))

[[  82    0    0    1    1  722   19    0    0    0    0   32]
 [   7    0    0    0    0  270    7    0    0    0    0   25]
 [   7    0    0    0    0   89    7    0    0    0    0    1]
 [   3    0    0    1    0  208    7    0    0    0    0   17]
 [  11    0    0    0    1  161    9    0    0    0    0    7]
 [  65    1    0    2    0 2459   31    0    0    0    0   66]
 [  43    0    0    1    0  895   76    0    0    0    0   30]
 [  11    0    0    0    0  286    7    0    0    0    0   26]
 [   2    0    0    0    0   38    0    0    0    0    0    1]
 [  11    0    1    0    0  119    3    0    0    1    0    1]
 [   5    0    0    0    0   36    1    0    0    0    0    4]
 [  33    0    0    0    0  934   17    1    0    0    0  135]]


In [96]:
# Print a classification report
print(metrics.classification_report(y_test,predictions))

              precision    recall  f1-score   support

      البسيط       0.29      0.10      0.14       857
      الخفيف       0.00      0.00      0.00       309
       الرجز       0.00      0.00      0.00       104
       الرمل       0.20      0.00      0.01       236
      السريع       0.50      0.01      0.01       189
      الطويل       0.40      0.94      0.56      2624
      الكامل       0.41      0.07      0.12      1045
    المتقارب       0.00      0.00      0.00       330
      المديد       0.00      0.00      0.00        41
     المنسرح       1.00      0.01      0.01       136
       الهزج       0.00      0.00      0.00        46
      الوافر       0.39      0.12      0.18      1120

    accuracy                           0.39      7037
   macro avg       0.27      0.10      0.09      7037
weighted avg       0.35      0.39      0.27      7037



C:\Users\snmuj\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\snmuj\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\snmuj\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [97]:
# Print the overall accuracy
print(metrics.accuracy_score(y_test,predictions))

0.39150206053716075


Naïve Bayes gave us better-than-average results at 76.4% for classifying reviews as positive or negative based on text alone. Let's see if we can do better.

## Feed the training data through the second pipeline
Next we'll run Linear SVC

In [98]:
text_clf_lsvc.fit(X_train, y_train)

Pipeline(steps=[('vect', CountVectorizer()), ('clf', LinearSVC())])

## Run predictions and analyze the results (Linear SVC)

In [99]:
# Form a prediction set
predictions = text_clf_lsvc.predict(X_test)

In [100]:
# Report the confusion matrix
from sklearn import metrics
print(metrics.confusion_matrix(y_test,predictions))

[[ 202   24    5   28   20  309  104   18    5   12    0  130]
 [  29   50    3    9   15   73   43   14    1    3    0   69]
 [   8    1    7    3    0   35   25    5    1    1    0   18]
 [  21    7    3   35    5   77   28    6    3    1    2   48]
 [  21    4    0    8   17   57   45    8    2    3    0   24]
 [ 193   59   16   42   34 1496  328   59    9   32    5  351]
 [  71   21   15   35   18  339  303   28    8    8    2  197]
 [  36   12    2    7    7   86   55   51    2    4    1   67]
 [   3    0    0    2    1   12   11    0    6    2    0    4]
 [  10    6    2    3    4   46   26    2    3   14    0   20]
 [   4    2    0    0    1    7    9    2    0    1    2   18]
 [  73   42    9   34   11  316  146   34    6    5    2  442]]


In [101]:
# Print a classification report
print(metrics.classification_report(y_test,predictions))

              precision    recall  f1-score   support

      البسيط       0.30      0.24      0.26       857
      الخفيف       0.22      0.16      0.19       309
       الرجز       0.11      0.07      0.08       104
       الرمل       0.17      0.15      0.16       236
      السريع       0.13      0.09      0.11       189
      الطويل       0.52      0.57      0.55      2624
      الكامل       0.27      0.29      0.28      1045
    المتقارب       0.22      0.15      0.18       330
      المديد       0.13      0.15      0.14        41
     المنسرح       0.16      0.10      0.13       136
       الهزج       0.14      0.04      0.07        46
      الوافر       0.32      0.39      0.35      1120

    accuracy                           0.37      7037
   macro avg       0.23      0.20      0.21      7037
weighted avg       0.36      0.37      0.36      7037



In [102]:
# Print the overall accuracy
print(metrics.accuracy_score(y_test,predictions))

0.3730282790962058
